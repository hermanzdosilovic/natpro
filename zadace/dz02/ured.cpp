#include <cstdio>
#include <algorithm>

using namespace std;

typedef long long lld;

const int MAXN = 100000;
const lld INF = 0x7fffffffffffffff;

int N, K, A, B;
char c[9];
lld tree[1 << 18], w[MAXN + 3], T;

void init(int n, int l, int r) {
  if (l == r) {
    tree[n] = w[l];
    return;
  }
  init(2*n, l, (l+r)/2);
  init(2*n + 1, (l+r)/2 + 1, r);
  tree[n] = min(tree[2*n], tree[2*n + 1]);
}

lld query(int n, int l, int r) {
  if (l > B || r < A) {
    return INF;
  } else if (l >= A && r <= B) {
    return tree[n];
  }
  lld q = query(2*n, l, (l+r)/2);
  lld p = query(2*n + 1, (l+r)/2 + 1, r);
  return min(q, p);
}

void update(int n, int l, int r) {
  if (l > A || r < A) {
    return;
  } else if (l == r && r == A) {
    tree[n] = w[A];
    return;
  } else if (l == r) {
    return;
  }
  update(2*n, l, (l+r)/2);
  update(2*n + 1, (l+r)/2 + 1, r);
  tree[n] = min(tree[2*n], tree[2*n + 1]);
}

int main() {
  scanf("%d", &N);
  for (int i = 1; i <= N; i++) {
    scanf("%lld", &w[i]);
  }
  init(1, 1, N);
  scanf("%d", &K);
  while (K--) {
    scanf("%s", c);
    if (c[0] == 's') {
      scanf("%d%d", &A, &B);
      printf("%lld\n", query(1, 1, N));
    } else {
      scanf("%d%lld", &A, &T);
      w[A] = T;
      update(1, 1, N);
    }
  }
  return 0;
}
