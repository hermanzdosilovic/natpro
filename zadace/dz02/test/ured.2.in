16
16 15 14 13 12 11 10 9 8 7 6 5 4 3 2 1
50
sef 4 15
promjena 2 13
promjena 15 3
promjena 10 2
sef 12 14
sef 4 7
sef 13 14
promjena 15 10
promjena 15 14
promjena 7 8
sef 15 16
promjena 14 9
sef 12 16
promjena 16 4
promjena 14 10
promjena 6 16
sef 5 11
promjena 7 6
sef 12 16
sef 10 15
promjena 12 15
promjena 1 11
promjena 15 1
promjena 7 14
promjena 10 3
promjena 15 14
sef 13 14
sef 5 6
promjena 1 7
promjena 12 5
promjena 8 3
sef 2 9
promjena 3 2
sef 2 10
sef 11 16
promjena 10 3
sef 3 7
sef 13 15
promjena 6 10
sef 10 11
promjena 11 10
promjena 4 16
sef 7 16
promjena 7 15
sef 13 15
promjena 4 16
sef 10 12
sef 10 14
sef 14 16
promjena 14 13
