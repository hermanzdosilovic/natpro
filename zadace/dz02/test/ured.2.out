2
3
10
3
1
1
6
4
4
4
12
8
8
4
12
4
6
4
4
15
4
5
